ergo-lib/src/wallet/box_selector/simple.rs

```diff
@@ -152,29 +166,144 @@ impl<T: ErgoBoxAssets> BoxSelector<T> for SimpleBoxSelector {
+pub struct CheckPreservationError(String);
+
+impl From<TokenAmountError> for CheckPreservationError {
+    fn from(e: TokenAmountError) -> Self {
+        CheckPreservationError(format!("TokenAmountError: {}", e))
+    }
+}
+
+/// Check if the selected inputs value and tokens are equal to the target + change
+fn check_input_preservation<T: ErgoBoxAssets>(
+    selected_inputs: &[T],
+    change_boxes: &[ErgoBoxAssetsData],
+    target_balance: BoxValue,
+    target_tokens: &[Token],
+) -> Result<(), CheckPreservationError> {
+    let sum_selected_inputs = sum_value(selected_inputs);
+    let sum_change_boxes = sum_value(change_boxes);
+    if sum_selected_inputs != sum_change_boxes + target_balance.as_u64() {
+        return Err(CheckPreservationError(
+            format!("total value of the selected boxes {:?} should equal target balance {:?} + total value in change boxes {:?}", sum_selected_inputs, target_balance.as_u64(), sum_change_boxes)
+        ));
+    }
+
+    let sum_tokens_selected_inputs = sum_tokens_from_boxes(selected_inputs)?;
+    let sum_tokens_change_boxes = sum_tokens_from_boxes(change_boxes)?;
+    let sum_tokens_target = sum_tokens(Some(target_tokens))?;
+    if sum_tokens_selected_inputs
+        != sum_tokens_from_hashmaps(sum_tokens_change_boxes.clone(), sum_tokens_target.clone())?
+    {
+        return Err(CheckPreservationError(
+            format!("all tokens from selected boxes {:?} should equal all tokens from the change boxes {:?} + target tokens {:?}", sum_tokens_selected_inputs, sum_tokens_change_boxes, sum_tokens_target)
+        ));
+    }
+    Ok(())
+}
+
+/// Not enough coins for change box(es)
+#[derive(Error, PartialEq, Eq, Debug, Clone)]
+#[error("Not enough coins for change box(es)")]
+pub struct NotEnoughCoinsForChangeBox(String);
+
+/// Split change tokens into a multiple boxes if over 255 distinct tokens
+fn make_change_boxes(
+    change_value: BoxValue,
+    change_tokens: HashMap<TokenId, TokenAmount>,
+) -> Result<Vec<ErgoBoxAssetsData>, NotEnoughCoinsForChangeBox> {
+    if change_tokens.is_empty() {
+        Ok(vec![ErgoBoxAssetsData {
+            value: change_value,
+            tokens: None,
+        }])
+    } else if change_tokens.len() < ErgoBox::MAX_TOKENS_COUNT {
+        #[allow(clippy::unwrap_used)]
+        // unwrap_used is ok here because we checked that change_tokens.len() < ErgoBox::MAX_TOKENS_COUNT
+        Ok(vec![ErgoBoxAssetsData {
+            value: change_value,
+            tokens: Some(
+                BoxTokens::from_vec(change_tokens.into_iter().map(Token::from).collect()).unwrap(),
+            ),
+        }])
+    } else {
+        let mut change_boxes = vec![];
+        let mut change_tokens_left: Vec<Token> =
+            change_tokens.into_iter().map(Token::from).collect();
+        let mut change_value_left = change_value;
+        while !change_tokens_left.is_empty() {
+            if change_tokens_left.len() < ErgoBox::MAX_TOKENS_COUNT {
```

Comment:
Nice spot.

Reply:
Good catch! Thank you! I guess I subconsciously tried to avoid the edge case (255 tokens). Fixed here and above(whole box, no split) and added an `ErgoBox` test with 255 tokens.